chat-android/src/main/java/com/ably/chat/AtomicCoroutineScope.kt

```diff
@@ -0,0 +1,93 @@
+package com.ably.chat
+
+import java.util.concurrent.PriorityBlockingQueue
+import kotlin.coroutines.cancellation.CancellationException
+import kotlin.coroutines.coroutineContext
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
+
+/**
+ * AtomicCoroutineScope is a thread safe wrapper to run multiple operations mutually exclusive.
+ * All operations are atomic and run with given priority.
+ * Accepts scope as a constructor parameter to run operations under the given scope.
+ * See [Kotlin Dispatchers](https://kt.academy/article/cc-dispatchers) for more information.
+ */
+class AtomicCoroutineScope(private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default)) {
+
+    private val sequentialScope: CoroutineScope = CoroutineScope(Dispatchers.Default.limitedParallelism(1))
```

Comment:
We do not want to cancel `sequentialScope`. Because, then we won't be able to run new jobs.
`cancel` is only supposed to cancel current and pending jobs as per method documentation

Reply:
Also, `sequentialScope` will be automatically garbage collected if no jobs are running as a part of it. You can search for related info. on the internet.

Reply:
Okay, then explain to me about garbage collection of sequential scope and how it will work : P